src/services/uniswap.js

```diff
@@ -0,0 +1,329 @@
+  const {
+    address: fromAddress, symbol: fromSymbol, name: fromName,
+  } = fromAsset;
+  const {
+    address: toAddress, symbol: toSymbol, name: toName,
+  } = toAsset;
+
+  try {
+    const token1 = await Token.fetchData(chainId, fromAddress, undefined, fromSymbol, fromName);
+    const token2 = await Token.fetchData(chainId, toAddress, undefined, toSymbol, toName);
+    const pair = await Pair.fetchData(token1, token2);
+
+    const route = new Route([pair], token1);
+    return route;
+  } catch (e) {
+    return getBackupRoute(fromAddress, toAddress);
+  }
+};
+
+const getTrade = async (
+  fromAssetAddress: string,
+  fromQuantityInBaseUnits: string,
+  route: Route,
+): Promise<Trade> => {
+  const fromToken = await Token.fetchData(chainId, toChecksumAddress(fromAssetAddress));
+  const fromTokenAmount = new TokenAmount(fromToken, fromQuantityInBaseUnits);
+  const trade = new Trade(route, fromTokenAmount, TradeType.EXACT_INPUT);
+  return trade;
+};
+
+const getAllowanceSet = async (clientAddress: string, fromAsset: Asset): Promise<boolean> => {
+  if (fromAsset.code === 'ETH') return true;
+  const assetContract = new ethers.Contract(fromAsset.address, ERC20_CONTRACT_ABI, ethProvider);
+  const allowance: BigNumber = await assetContract.allowance(clientAddress, ADDRESSES.router);
+  return allowance.gt(0);
+};
+
+export const getUniswapOffer = async (
+  fromAsset: Asset,
+  toAsset: Asset,
+  quantity: number | string,
+  clientAddress: string,
+): Promise<Offer | null> => {
+  parseAssets([fromAsset, toAsset]);
+  const decimalsBN = new BigNumber(fromAsset.decimals);
+  const quantityBN = new BigNumber(quantity);
+  const fromAssetQuantityBaseUnits = convertToBaseUnits(decimalsBN, quantityBN);
+  const route: ?Route = await getRoute(fromAsset, toAsset);
+  if (!route) return null;
+  const trade: Trade = await getTrade(fromAsset.address, fromAssetQuantityBaseUnits.toFixed(), route);
+  const askRate = getAskRate(trade);
+  const allowanceSet = await getAllowanceSet(clientAddress, fromAsset);
+  const offer: Offer = parseOffer(fromAsset, toAsset, allowanceSet, askRate, PROVIDER_UNISWAP);
+  return offer;
+};
+
+const getUniswapOrderData = async (
+  fromAsset: Asset,
+  toAsset: Asset,
+  fromAssetQuantityBaseUnits: string,
+  toAssetDecimals: string,
+): Promise<{ path: string[], expectedOutputBaseUnits: BigNumber }> => {
+  let route = await getRoute(fromAsset, toAsset);
+  if (!route && (
+    fromAsset.address.toLowerCase() === ADDRESSES.WETH.toLowerCase()
+      || toAsset.address.toLowerCase() === ADDRESSES.WETH.toLowerCase()
+  )) {
+    reportLog('Unable to find a possible route', null, 'error');
+  }
+  if (!route) {
+    route = await getBackupRoute(fromAsset.address, toAsset.address);
+    if (!route) {
+      reportLog('Unable to find a possible route', null, 'error');
+    }
+  }
+  const trade = await getTrade(fromAsset.address, fromAssetQuantityBaseUnits, route);
```

Comment:
All good to continue with the calls if no route? getTrade seems to accept only param for route which is not nullable.